File: Sources/SwiftMTH/MSDPTypes.swift
/// Flags for MSDP variable definitions and per-connection variable state.
///
/// Definition flags (set on the variable definition, shared across connections):
/// - `command`: This variable is a command (LIST, REPORT, SEND, etc.)
/// - `list`: This variable is a list category
/// - `sendable`: Value can be sent on demand
/// - `reportable`: Value can be auto-reported on change
/// - `configurable`: Value can be set by the client
///
/// Per-connection state flags (set per session):
/// - `reported`: Currently being auto-reported for this connection
/// - `updated`: Has changed since last flush
public struct MSDPFlags: OptionSet, Sendable {
    public let rawValue: Int

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    public static let command      = MSDPFlags(rawValue: 1 << 0)
    public static let list         = MSDPFlags(rawValue: 1 << 1)
    public static let sendable     = MSDPFlags(rawValue: 1 << 2)
    public static let reportable   = MSDPFlags(rawValue: 1 << 3)
    public static let configurable = MSDPFlags(rawValue: 1 << 4)
    public static let reported     = MSDPFlags(rawValue: 1 << 5)
    public static let updated      = MSDPFlags(rawValue: 1 << 6)
}

/// Definition of a single MSDP variable in the table.
public struct MSDPVariableDefinition: Sendable {
    public let name: String
    public let flags: MSDPFlags

    public init(name: String, flags: MSDPFlags) {
        self.name = name
        self.flags = flags
    }
}

/// Per-connection state for a single MSDP variable.
public struct MSDPVariableState {
    public var value: String
    public var flags: MSDPFlags

    public init(value: String = "", flags: MSDPFlags = []) {
        self.value = value
        self.flags = flags
    }
}

/// The default MSDP variable table matching the C `msdp_table[]`.
/// Alphabetically sorted as required by the MSDP specification.
public let defaultMSDPTable: [MSDPVariableDefinition] = [
    MSDPVariableDefinition(name: "ALIGNMENT",              flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "ARACHNOS_DEVEL",         flags: [.configurable, .reportable]),
    MSDPVariableDefinition(name: "ARACHNOS_MUDLIST",       flags: [.configurable]),
    MSDPVariableDefinition(name: "COMMANDS",               flags: [.command, .list]),
    MSDPVariableDefinition(name: "CONFIGURABLE_VARIABLES", flags: [.configurable, .list]),
    MSDPVariableDefinition(name: "EXPERIENCE",             flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "EXPERIENCE_MAX",         flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "HEALTH",                 flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "HEALTH_MAX",             flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "LEVEL",                  flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "LIST",                   flags: [.command]),
    MSDPVariableDefinition(name: "LISTS",                  flags: [.list]),
    MSDPVariableDefinition(name: "MANA",                   flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "MANA_MAX",               flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "MONEY",                  flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "MOVEMENT",               flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "MOVEMENT_MAX",           flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "REPORT",                 flags: [.command]),
    MSDPVariableDefinition(name: "REPORTABLE_VARIABLES",   flags: [.reportable, .list]),
    MSDPVariableDefinition(name: "REPORTED_VARIABLES",     flags: [.reported, .list]),
    MSDPVariableDefinition(name: "RESET",                  flags: [.command]),
    MSDPVariableDefinition(name: "ROOM",                   flags: [.reportable]),
    MSDPVariableDefinition(name: "ROOM_EXITS",             flags: [.sendable, .reportable]),
    MSDPVariableDefinition(name: "SEND",                   flags: [.command]),
    MSDPVariableDefinition(name: "SENDABLE_VARIABLES",     flags: [.sendable, .list]),
    MSDPVariableDefinition(name: "SPECIFICATION",          flags: [.sendable]),
    MSDPVariableDefinition(name: "UNREPORT",               flags: [.command]),
]

File: kotlin/mth-core/src/main/kotlin/mth/core/MSDPTypes.kt
package mth.core

@JvmInline
value class MSDPFlags(val rawValue: Int = 0) {
    operator fun contains(flag: MSDPFlags): Boolean = rawValue and flag.rawValue == flag.rawValue
    fun insert(flag: MSDPFlags) = MSDPFlags(rawValue or flag.rawValue)
    fun remove(flag: MSDPFlags) = MSDPFlags(rawValue and flag.rawValue.inv())
    fun subtract(flag: MSDPFlags) = MSDPFlags(rawValue and flag.rawValue.inv())
    fun isEmpty(): Boolean = rawValue == 0

    companion object {
        val COMMAND = MSDPFlags(1 shl 0)
        val LIST = MSDPFlags(1 shl 1)
        val SENDABLE = MSDPFlags(1 shl 2)
        val REPORTABLE = MSDPFlags(1 shl 3)
        val CONFIGURABLE = MSDPFlags(1 shl 4)
        val REPORTED = MSDPFlags(1 shl 5)
        val UPDATED = MSDPFlags(1 shl 6)

        infix fun Int.or(flag: MSDPFlags) = this or flag.rawValue
        fun of(a: MSDPFlags, b: MSDPFlags) = MSDPFlags(a.rawValue or b.rawValue)
    }
}

data class MSDPVariableDefinition(
    val name: String,
    val flags: MSDPFlags
)

class MSDPVariableState(
    var value: String = "",
    var flags: MSDPFlags = MSDPFlags()
)

val defaultMSDPTable: List<MSDPVariableDefinition> = listOf(
    MSDPVariableDefinition("ALIGNMENT", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("ARACHNOS_DEVEL", MSDPFlags.of(MSDPFlags.CONFIGURABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("ARACHNOS_MUDLIST", MSDPFlags(MSDPFlags.CONFIGURABLE.rawValue)),
    MSDPVariableDefinition("COMMANDS", MSDPFlags.of(MSDPFlags.COMMAND, MSDPFlags.LIST)),
    MSDPVariableDefinition("CONFIGURABLE_VARIABLES", MSDPFlags.of(MSDPFlags.CONFIGURABLE, MSDPFlags.LIST)),
    MSDPVariableDefinition("EXPERIENCE", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("EXPERIENCE_MAX", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("HEALTH", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("HEALTH_MAX", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("LEVEL", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("LIST", MSDPFlags(MSDPFlags.COMMAND.rawValue)),
    MSDPVariableDefinition("LISTS", MSDPFlags(MSDPFlags.LIST.rawValue)),
    MSDPVariableDefinition("MANA", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("MANA_MAX", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("MONEY", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("MOVEMENT", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("MOVEMENT_MAX", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("REPORT", MSDPFlags(MSDPFlags.COMMAND.rawValue)),
    MSDPVariableDefinition("REPORTABLE_VARIABLES", MSDPFlags.of(MSDPFlags.REPORTABLE, MSDPFlags.LIST)),
    MSDPVariableDefinition("REPORTED_VARIABLES", MSDPFlags.of(MSDPFlags.REPORTED, MSDPFlags.LIST)),
    MSDPVariableDefinition("RESET", MSDPFlags(MSDPFlags.COMMAND.rawValue)),
    MSDPVariableDefinition("ROOM", MSDPFlags(MSDPFlags.REPORTABLE.rawValue)),
    MSDPVariableDefinition("ROOM_EXITS", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.REPORTABLE)),
    MSDPVariableDefinition("SEND", MSDPFlags(MSDPFlags.COMMAND.rawValue)),
    MSDPVariableDefinition("SENDABLE_VARIABLES", MSDPFlags.of(MSDPFlags.SENDABLE, MSDPFlags.LIST)),
    MSDPVariableDefinition("SPECIFICATION", MSDPFlags(MSDPFlags.SENDABLE.rawValue)),
    MSDPVariableDefinition("UNREPORT", MSDPFlags(MSDPFlags.COMMAND.rawValue)),
)
